pyamg/amg_core/air.h

```diff
@@ -0,0 +1,584 @@
+#ifndef AIR_H
+#define AIR_H
+
+#include <iostream>
+#include <cmath>
+#include <vector>
+#include <iterator>
+#include <cassert>
+#include <limits>
+#include <algorithm>
+
+#include "linalg.h"
+#include "graph.h"
+#include "krylov.h"
+
+#define F_NODE 0
+#define C_NODE 1
+
+
+/* Interpolate C-points by value and each F-point by value from its strongest
+ * connected C-neighbor. 
+ * 
+ * Parameters
+ * ----------
+ *      Rp : const array<int> 
+ *          Pre-determined row-pointer for P in CSR format
+ *      Rj : array<int>
+ *          Empty array for column indices for P in CSR format
+ *      Cp : const array<int>
+ *          Row pointer for SOC matrix, C
+ *      Cj : const array<int>
+ *          Column indices for SOC matrix, C
+ *      Cx : const array<float>
+ *          Data array for SOC matrix, C
+ *      splitting : const array<int>
+ *          Boolean array with 1 denoting C-points and 0 F-points
+ *
+ * Returns
+ * -------
+ * Nothing, Rj[] modified in place.
+ *
+ */
+template<class I, class T>
+void one_point_interpolation(      I Pp[],    const int Pp_size,
+                                   I Pj[],   const int Pj_size,
+                                   T Px[],   const int Px_size,
+                             const I Cp[],  const int Cp_size,
+                             const I Cj[], const int Cj_size,
+                             const T Cx[],    const int Cx_size,
+                             const I splitting[], const int splitting_size)
+{
+    I n = Pp_size-1;
+
+    // Get enumeration of C-points, where if i is the jth C-point,
+    // then pointInd[i] = j.
+    std::vector<I> pointInd(n);
+    pointInd[0] = 0;
+    for (I i=1; i<n; i++) {
+        pointInd[i] = pointInd[i-1] + splitting[i-1];
+    }
+
+    Pp[0] = 0;
+    // Build interpolation operator as CSR matrix
+    I next = 0;
+    for (I row=0; row<n; row++) {
+
+        // Set C-point as identity
+        if (splitting[row] == C_NODE) {
+            Pj[next] = pointInd[row];
+            next += 1;
+        }
+        // For F-points, find strongest connection to C-point
+        // and interpolate directly from C-point. 
+        else {
+            T max = -1.0;
+            I ind = -1;
+            T val = 0.0;
+            for (I i=Cp[row]; i<Cp[row+1]; i++) {
+                if (splitting[Cj[i]] == C_NODE) {
+                    double vv = std::abs(Cx[i]);
+                    if (vv > max) {
+                        max = vv;
+                        ind = Cj[i];
+                        val = Cx[i];
+                    }
+                }
+            }
+            if (ind > -1) {
+              Pj[next] = pointInd[ind];
+              Px[next] = -val;
+              next += 1;
+            }
+        }
+        Pp[row+1] = next;
+    }
+}
+
+
+
+/* Sorting function for approx_ideal_restriction_pass1, doesn't like being
```

Comment:
@bensworth 
Is this function still used anywhere?  Maybe we can delete it, or figure out with Luke some way to template it?

Reply:
ya I dont know if we need it anymore. also dont remember why I had it in there. will remove